File: lab-06/scripts/lab-0606-analysis/comparison.js
// Модель данных
class User {
    constructor(id, name, email) {
        this.id = id;
        this.name = name;
        this.email = email;
    }
}

class Product {
    constructor(id, name, price, category) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.category = category;
    }
}

class OrderItem {
    constructor(product, quantity) {
        this.product = product;
        this.quantity = quantity;
    }
}

class Order {
    constructor(id, user, items, status) {
        this.id = id;
        this.user = user;
        this.items = items;
        this.status = status;
    }
}

// Пример данных
const users = [
    new User(1, "John Doe", "john@example.com"),
    new User(2, "Jane Smith", "jane@example.com")
];

const products = [
    new Product(1, "iPhone", 999.99, "electronics"),
    new Product(2, "MacBook", 1999.99, "electronics"),
    new Product(3, "T-shirt", 29.99, "clothing")
];

const orders = [
    new Order(1, users[0], [
        new OrderItem(products[0], 1),
        new OrderItem(products[2], 2)
    ], "completed"),
    new Order(2, users[1], [
        new OrderItem(products[1], 1)
    ], "pending")
];

// Функции обработки
const calculateOrderTotal = (order) => 
    order.items.reduce((total, item) => total + (item.product.price * item.quantity), 0);

const filterOrdersByStatus = (orders, status) => 
    orders.filter(order => order.status === status);

const getTopExpensiveOrders = (orders, n) => 
    [...orders].sort((a, b) => calculateOrderTotal(b) - calculateOrderTotal(a)).slice(0, n);

const applyDiscount = (order, discount) => {
    const discountedItems = order.items.map(item => ({
        ...item,
        product: {
            ...item.product,
            price: item.product.price * (1 - discount)
        }
    }));
    return { ...order, items: discountedItems };
};

// ===== ✓ ПРАКТИЧЕСКИЕ ЗАДАНИЯ =====
// Задание 1: Бенчмарк
console.time("JS Benchmark");
for(let i = 0; i < 1000; i++) {
    orders.filter(o => o.status === "completed").reduce((s, o) => s + calculateOrderTotal(o), 0);
}
console.timeEnd("JS Benchmark");  // ✓ Задание 1

// Основная функция
const main = () => {
    const completedOrders = filterOrdersByStatus(orders, "completed");
    const totalRevenue = completedOrders.reduce((sum, order) => sum + calculateOrderTotal(order), 0);
    const topOrders = getTopExpensiveOrders(orders, 2);
    
    console.log("Общая выручка:", totalRevenue);
    console.log("Топ заказы:", topOrders.map(calculateOrderTotal));
};

main();

File: lab-06/scripts/lab-0606-analysis/comparison.rs
#[derive(Debug, Clone)]
struct User {
    id: u32,
    name: String,
    email: String,
}

#[derive(Debug, Clone)]
struct Product {
    id: u32,
    name: String,
    price: f64,
    category: String,
}

#[derive(Debug, Clone)]
struct OrderItem {
    product: Product,
    quantity: u32,
}

#[derive(Debug, Clone)]
struct Order {
    id: u32,
    user: User,
    items: Vec<OrderItem>,
    status: String,
}

impl User {
    fn new(id: u32, name: &str, email: &str) -> Self {
        User {
            id,
            name: name.to_string(),
            email: email.to_string(),
        }
    }
}

impl Product {
    fn new(id: u32, name: &str, price: f64, category: &str) -> Self {
        Product {
            id,
            name: name.to_string(),
            price,
            category: category.to_string(),
        }
    }
}

impl OrderItem {
    fn new(product: Product, quantity: u32) -> Self {
        OrderItem { product, quantity }
    }
}

impl Order {
    fn new(id: u32, user: User, items: Vec<OrderItem>, status: &str) -> Self {
        Order {
            id,
            user,
            items,
            status: status.to_string(),
        }
    }
}

// Функции обработки
fn calculate_order_total(order: &Order) -> f64 {
    order.items.iter()
        .map(|item| item.product.price * item.quantity as f64)
        .sum()
}

fn filter_orders_by_status(orders: &[Order], status: &str) -> Vec<Order> {
    orders.iter()
        .filter(|order| order.status == status)
        .cloned()
        .collect()
}

fn get_top_expensive_orders(orders: &[Order], n: usize) -> Vec<Order> {
    let mut sorted_orders = orders.to_vec();
    sorted_orders.sort_by(|a, b| {
        calculate_order_total(b).partial_cmp(&calculate_order_total(a)).unwrap()
    });
    sorted_orders.into_iter().take(n).collect()
}

fn apply_discount(order: &Order, discount: f64) -> Order {
    let discounted_items: Vec<OrderItem> = order.items.iter()
        .map(|item| {
            let discounted_product = Product {
                price: item.product.price * (1.0 - discount),
                ..item.product.clone()
            };
            OrderItem {
                product: discounted_product,
                ..item.clone()
            }
        })
        .collect();
    
    Order {
        items: discounted_items,
        ..order.clone()
    }
}

// ===== ✓ ПРАКТИЧЕСКИЕ ЗАДАНИЯ =====
// Задание 1: Бенчмарк
fn benchmark_rust(orders: &[Order]) {
    let start = std::time::Instant::now();
    for _ in 0..1000 {
        let _ = orders.iter().filter(|o| o.status == "completed").map(calculate_order_total).sum::<f64>();
    }
    println!("Rust ✓ Бенчмарк (1000ит): {}ms", start.elapsed().as_millis());
}

// Задание 2: Анализ безопасности типов (ТОЧНО ИЗ УСЛОВИЯ)
fn type_safe_calculation(order: &Order) -> f64 {
    // Компилятор проверит все типы
    order.items.iter()
        .map(|item| item.product.price * f64::from(item.quantity))
        .sum() // Тип возвращаемого значения проверяется
}

fn main() {
    let users = vec![
        User::new(1, "John Doe", "john@example.com"),
        User::new(2, "Jane Smith", "jane@example.com"),
    ];
    
    let products = vec![
        Product::new(1, "iPhone", 999.99, "electronics"),
        Product::new(2, "MacBook", 1999.99, "electronics"),
        Product::new(3, "T-shirt", 29.99, "clothing"),
    ];
    
    let orders = vec![
        Order::new(
            1, 
            users[0].clone(), 
            vec![
                OrderItem::new(products[0].clone(), 1),
                OrderItem::new(products[2].clone(), 2)
            ], 
            "completed"
        ),
        Order::new(
            2, 
            users[1].clone(), 
            vec![
                OrderItem::new(products[1].clone(), 1)
            ], 
            "pending"
        ),
    ];
    
    // ✓ Практические задания
    benchmark_rust(&orders);  // Задание 1
    println!("Rust ✓ type_safe_calculation: {}", type_safe_calculation(&orders[0]));  // Задание 2
    
    let completed_orders = filter_orders_by_status(&orders, "completed");
    let total_revenue: f64 = completed_orders.iter().map(calculate_order_total).sum();
    let top_orders = get_top_expensive_orders(&orders, 2);
    
    println!("Общая выручка: {:.2}", total_revenue);
    println!("Топ заказы: {:?}", top_orders.iter().map(calculate_order_total).collect::<Vec<f64>>());
}
